### PR TITLE
Remove LXC support.

### PR DESCRIPTION
**Pull request by calavera**
The LXC driver was deprecated in Docker 1.8.
Following the deprecation rules, we can remove a deprecated feature
after two major releases. LXC won't be supported anymore starting on Docker 1.10.

It also removes the global daemon option `--exec-driver` because it doesn't make sense anymore.

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by tianon** [comment]
LGTM! :+1:

**Comment by thaJeztah** [comment]
Think there's also some docs around lxc here https://github.com/docker/docker/blob/master/docs/reference/run.md#runtime-privilege-linux-capabilities-and-lxc-configuration (but we're not in docs review yet)

**Comment by cpuguy83** [comment]
runconfig/parse.go:65: flLxcOpts declared and not used

**Comment by calavera** [comment]
thanks @cpuguy83 

@thaJeztah yeah, I left a few things behind. I wanted to check if it actually worked :)
There are a few things in contrib that I don't really know how to clean, apparmor, selinux and other related scripts.

**Comment by thaJeztah** [comment]
@calavera no problem, just mentioned it so that we don't skip docs review on this one. I can help checking once we're there

**Comment by calavera** [comment]
@docker/maintainers please, make sure you check this PR, because it's happening :grin:

File: Dockerfile [hunk 1]
```diff
@@ -72,16 +72,6 @@ RUN cd /usr/local/lvm2 \
 	&& make install_device-mapper
 # see https://git.fedorahosted.org/cgit/lvm2.git/tree/INSTALL
 
-# Install lxc
```
Comment (tiborvass) [inline Dockerfile]: ugh i know it's silly but if we could merge this at the same time as #17701 to suffer only one cache bust :)

**Comment by vieux** [comment]
@calavera are we going to keep the execdrivers since there is only one now ?
I think it's a good idea if one another ever pops up.

**Comment by calavera** [comment]
@vieux that's a very good question. Maybe we can just remove that option all together. After all, windows and linux are already separated by build flags.

File: Dockerfile (hunk 1 above)
Comment (calavera) [inline Dockerfile]: we can put both changes in the same PR to merge them at the same time, although it's unusual.

**Comment by calavera** [comment]
I've also removed the global daemon option `--exec-driver` in a separated commit. Because as @vieux mentions, it doesn't make sense to keep it anymore.

**Comment by rhatdan** [comment]
Does this mean dockerinit goes away?  Can we switch to building docker as dynbinary as default?

**Comment by vdemeester** [comment]
@calavera the `--exec-driver` flag will go without deprecation notice though, right ? (I don't think this is a bad thing but prefer to mention/ask about it).

**Comment by cpuguy83** [comment]
Yeah, thought abut this too. Ultimately I think it should fail and not just silently ignore the fact that the user was specifying LXC.

**Comment by calavera** [comment]
@rhatdan dynbinary has been the default since 1.8.

@vdemeester yes it goes away without deprecation notice. It doesn't make sense to keep it as deprecated when it doesn't do anything.

@cpuguy83 it fails if you set the flag, because the parser doesn't recognize that option :smile:

**Comment by vdemeester** [comment]
@calavera ok,seems fair. LGTM then :-)

**Comment by ewindisch** [comment]
Conceptually, LGTM! Only glanced at the change itself, but nothing shocking.

**Comment by tiborvass** [comment]
I'm not sure about removing `--exec-driver` even if there's just now one driver left. In fact there are two drivers left, one windows and libcontainer even if you can't really choose between them because they are for different platforms.

Still, removing `--exec-driver` is dependent on the runc integration, since runc integration is what will give people a way to write and choose exec drivers. But those decisions are not finalized. If today, we remove the flag only to add it back later, it doesn't make sense. I prefer to live with a nop flag then with back and forth decisions.

To complicate the matter further, I'd be interested in moving the `--exec-driver` flag to the `run` command so that i can have multiple exec drivers at the same time.

**Comment by calavera** [comment]
> To complicate the matter further, I'd be interested in moving the --exec-driver flag to the run command so that i can have multiple exec drivers at the same time.

@tiborvass That's one more reason to remove this flag from the daemon. It's a completely different feature.

I'd rather not keep flags that we don't use only just in case. It will be easier to move forward with runc if we start with a clean slate and we don't try to keep the old idioms around only because we had them already.

@crosbymichael I'd really appreciate your input in this.

**Comment by tiborvass** [comment]
>  It will be easier to move forward with runc if we start with a clean slate and we don't try to keep the old idioms around only because we had them already.

@calavera that's why it's the runc integration PR that should remove the flag.

Anyway I won't fight it.

**Comment by LK4D4** [comment]
Btw, libnetwork depends on `.dockerinit` inside container.

**Comment by cpuguy83** [comment]
noooo

**Comment by jfrazelle** [comment]
hahaha that could be a phase 2?

**Comment by calavera** [comment]
@LK4D4 it looks like `.dockerinit` still exists in a container booted with this branch, I'm not sure if this changes anything:

```
➜  docker git:(remove_lxc) bundles/1.10.0-dev/binary/docker run --rm -it alpine ash
/ # ls -la
total 56
drwxr-xr-x   17 root     root          4096 Nov  5 21:15 .
drwxr-xr-x   17 root     root          4096 Nov  5 21:15 ..
-rwxr-xr-x    1 root     root             0 Nov  5 21:15 .dockerenv
-rwxr-xr-x    1 root     root             0 Nov  5 21:15 .dockerinit
drwxr-xr-x    2 root     root          4096 Sep 14 19:17 bin
```

It looks like libnetwork only uses that file to decide whether to raise an error or not. I'm not sure if we need to do anything, but if we do, maybe it's something that we can do in some other way that doesn't block this PR:

https://github.com/docker/libnetwork/blob/9fb7ba8fa01f0ddce99713799e760223c842fb4b/drivers/bridge/setup_bridgenetfiltering.go#L36-L41

**Comment by tianon** [comment]
AFAIUI, this PR technically opens the possibility of ditching dockerinit,
but IMO that discussion should be separate from this PR.

**Comment by calavera** [comment]
I agree with @tianon

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
lets move ahead!

**Comment by thaJeztah** [comment]
@calavera impact/deprecation too? It's marked deprecated already, but for easier finding.. Idk

**Comment by calavera** [comment]
@thaJeztah impact/:skull: maybe? :grin:

**Comment by vdemeester** [comment]
impact/:ghost:

**Comment by thaJeztah** [comment]
`impact/byebye-it-was-nice-knowing-you-but-we-moved-on`

**Comment by jfrazelle** [comment]
impact/RIP-jess-is-glad-for-after-over-2-years-never-having-to-run-lxc-start-again
:innocent:

**Comment by thaJeztah** [comment]
`impact deprecate !rebuild docker#17700/lxc`

**Comment by shykes** [comment]
LGTM

**Comment by michaelneale** [comment]
Shed a single tear

**Comment by jhowardmsft** [comment]
Ah please leave --exec-driver for the moment. There will soon be two Windows drivers while I'm developing the libcontainer Windows one

**Comment by xnox** [comment]
How is OCI exec driver going to be added?

Note that native driver / libcontainer != OCI. There are three OCI complaint executors already: clr, runv, runc. With only one of them based on libcontainer bits.

This kills clr exec-driver that uses VMs on Linux, which Highland team was about to be piloting to use. And which was requested to be merged to master...

Given that imminently we will need oci exec driver to supplement and migrate from native driver on linux at least, removing this support adds no value, and makes current and future development match harder.

Please revert?

**Comment by stevvooe** [comment]
@xnox Might be better to open a separate issue rather than commenting on a closed pull request. I would also suggest you be more specific about why this affects OCI integration.

**Comment by kimh** [comment]
I just want to be clear, but does this PR mean that I cannot run Docker on LXC from docker 1.10? Or is there a way other than using `--exec-driver lxc` to do this?

**Comment by thaJeztah** [comment]
@kimh correct; are there specific reasons you cannot use the "native" driver?

**Comment by kimh** [comment]
@thaJeztah 

I work in CircleCI and we use LXC container to run builds. Therefore, we've used LXC exec-driver to run Docker.

>are there specific reasons you cannot use the "native" driver?

Actually, I'm wondering if I can use libcontainer on LXC and made this [SO](http://stackoverflow.com/questions/34147405/using-libcontainer-on-lxc) question which led me to this PR.

So repeating the same question, is it technically possible to use libcontainer driver on LXC container?

**Comment by cpuguy83** [comment]
@kimh There should be nothing stopping it.

**Pull request merged**